cql3/statements/alter_keyspace_statement.cc

```diff
@@ -142,11 +199,18 @@ cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_proce
         auto ks_md_update = _attrs->as_ks_metadata_update(ks_md, tm, feat);
         std::vector<mutation> muts;
         std::vector<sstring> warnings;
-        auto ks_options = _attrs->get_all_options_flattened(feat);
+        bool include_tablet_options = _attrs->get_map(_attrs->KW_TABLETS).has_value();
```

Comment:
> I don't understand. If KW_TABLTETS is missing from the new map (_attrs), you will pass false to get_old_options_flattened as well, which means you will not include the tablets options (initial_tablets for example) at all to the final ks_options -- even the old ones. This contradicts what you said in the commit message, that we want to preserve the old options.

Not specifying tablets will result in leaving the old ones. However it works differently for other options - if they are not specified, they're set to the defaults. 
I could specify tablets options, to not create such special cases, but I lack data to handle all the cases here (how where they set, for exampe, via the scylla config or cql), so I'm doing this "trick" of not setting nothing.

> You also mentioned in the commit message that test_alter_preserves_tablets_if_initial_tablets_skipped test is modified to catch this, but the commit contains no such change to this test.

Sorry, I removed this part. IDK what happened, apparently this change is no longer needed, and most likely has been lost in rebases and reworks.

Reply:
Resolving. Feel free to reopen if needed.

Reply:
> Not specifying tablets will result in leaving the old ones. 

Then please explain where my misunderstanding of the code is. Because my analysis of the code:

> If KW_TABLTETS is missing from the new map (_attrs), you will pass false to get_old_options_flattened as well, which means you will not include the tablets options (initial_tablets for example) at all to the final ks_options -- even the old ones. 

shows that the opposite of what you claim is happening: i.e.
- if you don't specify tablets, we **don't** take the old options
- if you do specify tablets, we **do take** (merge with) the old options

Therefore I must've misunderstood the code.

Reply:
> Sorry, I removed this part. IDK what happened, apparently this change is no longer needed, and most likely has been lost in rebases and reworks.

Checking this behavior with a test is definitely something we should do, so if there is a test that can be adjusted, please adjust it, if not, add a new one

Reply:
> Then please explain where my misunderstanding of the code is

This code:
```
        bool include_tablet_options = _attrs->get_map(_attrs->KW_TABLETS).has_value();
        auto old_ks_options = get_old_options_flattened(ks, include_tablet_options);
        auto ks_options = get_current_options_flattened(_attrs, include_tablet_options, feat);
```
has the following effect:
1. if tablets are specified via CQL -> we're inserting tablets options into the map, thus in the end ALTERing tablets in the KS according to the new tablets parameters,
2. if tablets are not specified via CQL -> we're not placing any tablets parameter into the ks options map, thus not modifying tablets in the KS in the end - they stay untouched.

Why we need to repeat all options (rf, storage, durable writes) but not tablets - IDK, I suspect that later in the processing path something is different for processing tablets options, which requires this change here. 

> Checking this behavior with a test is definitely something we should do, so if there is a test that can be adjusted, please adjust it, if not, add a new one

This change is actually asserted by the tests, which check that we correctly set tablets via ALTER and correctly preserve tablets options via ALTER (`test_alter_preserves_tablets_if_initial_tablets_skipped`). What I originally had in mind was that this testcase I mentioned failed and I thought it's because it doesn't wait for ALTER tablets KS to complete, but I no longer see this, so I assume that my implementation at that time was not fully working.

Reply:
> Why we need to repeat all options (rf, storage, durable writes) but not tablets - IDK, I suspect that later in the processing path something is different for processing tablets options, which requires this change here.

This may be (related to) https://github.com/scylladb/scylladb/issues/20768

Reply:
So we determined that including old values of options such as storage might actually revert them to an older value due to https://github.com/scylladb/scylladb/issues/20768. Because the old values are taken from in-memory keyspace_metadata, which is not refreshed from disk on ALTER. We don't see that only because nothing is testing those options.